### ISSUE TITLE
Suggestion to add a survey-specific n() function

### ISSUE THREAD

**Issue by ray-p144** [gergness/srvyr#136]
I may have missed it, but I could not find a function to count the raw number of observations that are contributing to a certain analysis (i.e., the count of observations with non-zero weights).

A function like this would be useful when analyzing surveys like NHANES where multiple weights are provided for different subsets. The survey includes a `wtint2yr` interview weight and a `wtmec2yr` MEC exam weight. Those who did not participate in the MEC exam were given a zero weight for `wtmec2yr`. If the user was using the `wtmec2yr` weight and wanted to obtain a count of the number of observations that are contributing to the analysis, the `unweighted(n())` method (and the regular `n()` method) would still include all those with a weight of zero.

``` r
set.seed(1)

library(survey)
#> Loading required package: grid
#> Loading required package: Matrix
#> Loading required package: survival
#> 
#> Attaching package: 'survey'
#> The following object is masked from 'package:graphics':
#> 
#>     dotchart
library(srvyr)
#> 
#> Attaching package: 'srvyr'
#> The following object is masked from 'package:stats':
#> 
#>     filter
data(api)

apistrat %>%
    filter(pw == 0) %>%
    NROW()
#> [1] 0

apistrat <- apistrat %>%
    mutate(pw = ifelse(runif(n()) < .50, 0, pw))

apistrat %>%
    filter(pw == 0) %>%
    NROW()
#> [1] 98

strat_design_srvyr <- apistrat %>%
    as_survey_design(1, strata = stype, fpc = fpc, weight = pw)

strat_design_srvyr %>%
    summarise(count = unweighted(n()))
#> # A tibble: 1 x 1
#>   count
#>   <int>
#> 1   200

strat_design_srvyr %>%
    summarise(count = n())
#> # A tibble: 1 x 1
#>   count
#>   <int>
#> 1   200

survey_raw_n <- function () {
    .svy <- cur_svy()
    pweights <- (1 / .svy$prob)
    n <- sum(pweights > 0)
    return(n)
}

strat_design_srvyr %>%
    summarise(count = survey_raw_n())
#> # A tibble: 1 x 1
#>   count
#>   <int>
#> 1   102

sessionInfo()
#> R version 4.1.1 (2021-08-10)
#> Platform: x86_64-w64-mingw32/x64 (64-bit)
#> Running under: Windows 10 x64 (build 19044)
#> 
#> Matrix products: default
#> 
#> locale:
#> [1] LC_COLLATE=English_United States.1252 
#> [2] LC_CTYPE=English_United States.1252   
#> [3] LC_MONETARY=English_United States.1252
#> [4] LC_NUMERIC=C                          
#> [5] LC_TIME=English_United States.1252    
#> 
#> attached base packages:
#> [1] grid      stats     graphics  grDevices utils     datasets  methods  
#> [8] base     
#> 
#> other attached packages:
#> [1] srvyr_1.1.0     survey_4.1-1    survival_3.2-11 Matrix_1.3-4   
#> 
#> loaded via a namespace (and not attached):
#>  [1] compiler_4.1.1    pillar_1.6.2      highr_0.9         tools_4.1.1      
#>  [5] digest_0.6.27     evaluate_0.14     lifecycle_1.0.0   tibble_3.1.3     
#>  [9] lattice_0.20-44   pkgconfig_2.0.3   rlang_0.4.11      reprex_2.0.1     
#> [13] DBI_1.1.1         cli_3.1.0         rstudioapi_0.13   yaml_2.2.1       
#> [17] xfun_0.25         withr_2.4.2       stringr_1.4.0     dplyr_1.0.7      
#> [21] knitr_1.33        generics_0.1.0    fs_1.5.0          vctrs_0.3.8      
#> [25] mitools_2.4       tidyselect_1.1.1  glue_1.4.2        R6_2.5.0         
#> [29] fansi_0.5.0       rmarkdown_2.10    tidyr_1.1.3       purrr_0.3.4      
#> [33] magrittr_2.0.1    htmltools_0.5.1.1 ellipsis_0.3.2    splines_4.1.1    
#> [37] assertthat_0.2.1  utf8_1.2.2        stringi_1.7.3     crayon_1.4.1
```

<sup>Created on 2021-12-12 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by gergness** [comment]
Hm, maybe we need to get rid of this line?
https://github.com/gergness/srvyr/blob/main/R/survey_statistics.r#L759

@bschneidr - any thoughts? Do you think `unewighted()` should always remove rows where the probability is Inf? Currently it checks if `is.calibrated()` or `is.pps()`, I assume because those are the times when the survey package leaves observations in the dataset after subsetting. This issue suggests (at least someone) expects `unweighted()` to drop any row with infinite probability.

**Comment by ray-p144** [comment]
@gergness just to clarify my suggestion a little bit:

I would not have expected `unweighted()` to drop the rows with infinite probability. I actually would have expected to leave them in as it did because I would assume that it would not take the weights into account when counting the rows. I included it in my suggestion just to show that there was no method for counting the number of observations with a finite probability (and therefore are actually included in the other weighted analyses, like `survey_mean()`, `survey_quantile()`, etc.).

Personally, I think adding a separate function would be better than altering the `unweighted()` function.

**Comment by bschneidr** [comment]
This is a bit tricky. Like @ray-p144 said, it seems weird for `unweighted()` to be affected by the weights (i.e. by `1/my_design$prob`). 

@gerness you're right that the reason we have special behavior only for calibrated/PPS designs is that for other designs, subsetting/filtering the design drops the filtered-out rows from `my_design$variables` whereas for calibrated/PPS designs subsetting simply sets the probabilities equal to zero. For calibrated/raked designs, `unweighted()` doesn't distinguish between zero-weight rows caused by filtering vs. the user creating weights of zero in some other way (whether that's by supplying them from the start as @ray-p144 does, or perhaps as a result of, say, non-response adjustments).

For the problem that @ray-p144 is trying to solve here (count the number of rows with nonzero weights), here are the two solutions I'd recommend with the current functionality of `srvyr`:

1. Access the weights using `cur_svy()$prob`. 

```r
strat_design_srvyr %>%
  summarize(n_positive_wts = sum( (1/cur_svy()$prob) > 0))
#> # A tibble: 1 x 1
#> n_positive_wts
#>                <int>
#> 1                98
```

2. Access the original `pw` variable from the data, which is included in the design object's variables.
```r
strat_design_srvyr <- apistrat %>%
  as_survey_design(1, strata = stype, fpc = fpc, weight = pw)

"pw" %in% colnames(strat_design_srvyr)
#> [1] TRUE

strat_design_srvyr %>%
    summarize(n_positive_orig_wts = unweighted(sum(pw > 0)))

#> # A tibble: 1 x 1
#> n_positive_orig_wts
#>                <int>
#> 1                98
```

I don't think it's worth adding a new function just to count non-zero weights. However, it might be worth adding a helper function like `svy_wts()` which acts as a shorthand for `1/cur_svy()$prob`. I could see that being useful, especially since I think many users don't really understand how the weights/probabilities are stored and managed with the survey package.

**Comment by ray-p144** [comment]
@bschneidr I agree with the idea to add a helper function like `svy_wts()`. I actually like it better than my idea of adding a function to count non-zero weights.

**Comment by gergness** [comment]
Thanks Ben, yeah, I was struggling with having something so similar to `unweighted()`. I like the idea of adding (slightly renamed) `cur_svy_wts()`.

**Comment by gergness** [comment]
closed via #139
